### DISCUSSION TITLE
VRP: trucks starts from different lat/lon and ends with different locations

### DISCUSSION THREAD

**Discussion by Westlife1002** [google/or-tools#2620] [Routing (and legacy CP) questions]
I have developed a VRP model based on ortools/vrp package. the full code is here:
https://gist.github.com/Westlife1002/d920226616951d2b14d91e25eb9f23a1

1) Question 1
previously the model only considers all the trucks starting/ending from one unique warehouse location. A step further, I would like to define the truck starts from an arbitary GPS position. The reason is that we obtain the truck location based on the real-time GPS location data. So the truck is not necessarily standby at a given warehouse or customer location, and this means there are certain empty mils for trucks to go to the loading/unloading addresses.

I have seen this solution: https://developers.google.com/optimization/routing/routing_tasks#setting-start-and-end-locations-for-routes  however, this assumes trucks must start from a named location in the model (customer sites or warehouse)
The needs here is to initialize the trucks at an arbitary location. How to add this function in the model?

2) Question 2
above description asssumes that we know in advance how many customer demands and how many available trucks (incl. their types, number and locations) we can use for scheduling the goods pickup. However, if we only know the customer demands in advance, what is the best way for us to set the trucks' starts/ends locations in order to generate a proper solution. We then can use this solution to order trucks from the market.

**Comment by Mizux** [comment]
1. Technically solver only need a list of location indexed from 0 to N.
So you can adapt your model matrix to add the position of the vehicle as a location and use the overload with start/end to set the start position of each vehicle to be its actual position...

2. You could add a new vehicle whose "virtual" start position is N time away of any point to simulate the time the market needs to dispatch this new truck to any location requested...

**Reply by Westlife1002** [reply]
> 1. Technically solver only need a list of location indexed from 0 to N.
>    So you can adapt your model matrix to add the position of the vehicle as a location and use the overload with start/end to set the start position of each vehicle to be its actual position...
> 2. You could add a new vehicle whose "virtual" start position is N time away of any point to simulate the time the market needs to dispatch this new truck to any location requested...

thanks Mizux, I have tried to add two arrays of starts and ends points into the manager. Then I got below problem:

```
Traceback (most recent call last):
  File "C:\Users\user\AppData\Local\Programs\Python\Python36\lib\tkinter\__init__.py", line 1702, in __call__
    return self.func(*args)
  File "C:/Users/user/Dropbox/user/VRP_11.0.py", line 713, in main
    add_time_window_constraints(routing, data, time_callback_index, manager)
  File "C:/Users/user/Dropbox/user/VRP_11.0.py", line 397, in add_time_window_constraints
    time_dimension.CumulVar(index).SetRange(time_window[0], time_window[1])
  File "C:\Users\user\AppData\Roaming\Python\Python36\site-packages\ortools\constraint_solver\pywrapcp.py", line 1098, in SetRange
    return _pywrapcp.IntExpr_SetRange(self, l, u)
Exception: CP Solver fail
```
I could not understand fully your answer to the above two questions. Would you mind to help edit the code in my gists for me to better understand? 
full code: https://gist.github.com/Westlife1002/d920226616951d2b14d91e25eb9f23a1
